File: 1339.py
class Solution:
    def __init__(self):
        self.subtreeSum = list()

    def calSubtreeSum(self, node: Optional[TreeNode]) -> int:
        if node is None:
            return 0
        v = self.calSubtreeSum(node.left) + self.calSubtreeSum(node.right) + node.val
        self.subtreeSum.append(v)
        return v

    def maxProduct(self, root: Optional[TreeNode]) -> int:
        totalSum = self.calSubtreeSum(root)
        halfSum = totalSum // 2
        closestVal = None
        closestDist = None
        for s in self.subtreeSum:
            if closestDist is None or abs(s - halfSum) < closestDist:
                closestDist = abs(s - halfSum)
                closestVal = s
                if closestVal == halfSum: # not sure if this really speeds up
                    break
        return (closestVal * (totalSum - closestVal)) % (int(1e9) + 7)

File: 1339.cpp
struct Solution {
    static constexpr unsigned long m = 1000000007;
    vector<unsigned long> subsum; // surprisingly using forward_list yield error on first testcase

    unsigned long calSubSum(const TreeNode* root){
        if (root == nullptr){
            return 0;
        }
        unsigned long v = root->val + this->calSubSum(root->left) + this->calSubSum(root->right);
        this->subsum.push_back(v);
        return v;
    }

    int maxProduct(const TreeNode* root) {
        unsigned long totalSum = this->calSubSum(root);
        unsigned long halfSum = totalSum / 2, closestVal, closestDist, dist, v;
        bool flag = false;
        for (vector<unsigned long>::const_iterator it = this->subsum.cbegin();
             it != this->subsum.cend();
             ++it){
            v = *it;
            dist = (v >= halfSum) ? (v - halfSum) : (halfSum - v);
            if (!flag || dist < closestDist){
                flag = true;
                closestDist = dist;
                closestVal = v % this->m;
            }
        }
        return (closestVal * ((totalSum % this->m) - closestVal)) % this->m;
    }
};
